# CASE STUDY 1 (ARTIFICIAL INTELIGENCE on the production)

# Name :  ARBAJ SIKANDAR SHAIKH
    

## GFIPL 18052644

### Reduce the time spends on the production .

To ensure the safety and reliability of every unique configuration before they hit the consumer, the company’s engineers have developed a robust testing system. As one of the india’s biggest manufacturers of premium oil, safety and efficiency are paramount on GFIPL’s production lines. However, optimizing the speed of their production system for many possible feature combinations is complex and time-consuming without a powerful algorithmic approach.
You are required to reduce the time that spend on the production. Others will work with a dataset representing different permutations of features in a GFIPL to predict the time it takes to pass production. Optimal algorithms will contribute to faster production.


In [1]:
import pandas as pd
X=pd.read_csv(r"C:\Users\Arbaj\Downloads\train\train.csv")
Y=pd.read_csv(r"C:\Users\Arbaj\Downloads\test\test.csv")
X

ID       y  X0 X1  X2 X3 X4  X5 X6 X8  ...  X375  X376  X377  X378  \
0        0  130.81   k  v  at  a  d   u  j  o  ...     0     0     1     0   
1        6   88.53   k  t  av  e  d   y  l  o  ...     1     0     0     0   
2        7   76.26  az  w   n  c  d   x  j  x  ...     0     0     0     0   
3        9   80.62  az  t   n  f  d   x  l  e  ...     0     0     0     0   
4       13   78.02  az  v   n  f  d   h  d  n  ...     0     0     0     0   
...    ...     ...  .. ..  .. .. ..  .. .. ..  ...   ...   ...   ...   ...   
4204  8405  107.39  ak  s  as  c  d  aa  d  q  ...     1     0     0     0   
4205  8406  108.77   j  o   t  d  d  aa  h  h  ...     0     1     0     0   
4206  8412  109.22  ak  v   r  a  d  aa  g  e  ...     0     0     1     0   
4207  8415   87.48  al  r   e  f  d  aa  l  u  ...     0     0     0     0   
4208  8417  110.85   z  r  ae  c  d  aa  g  w  ...     1     0     0     0   

      X379  X380  X382  X383  X384  X385  
0        0     0     0     0     0     0  
1        0     0     0     0     0     0  
2        0     0     1     0     0     0  
3        0     0     0     0     0     0  
4        0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...  
4204     0     0     0     0     0     0  
4205     0     0     0     0     0     0  
4206     0     0     0     0     0     0  
4207     0     0     0     0     0     0  
4208     0     0     0     0     0     0  

[4209 rows x 378 columns]

In [2]:
y=X[['y']]
y

y
0     130.81
1      88.53
2      76.26
3      80.62
4      78.02
...      ...
4204  107.39
4205  108.77
4206  109.22
4207   87.48
4208  110.85

[4209 rows x 1 columns]

In [3]:
Y

ID  X0  X1  X2 X3 X4  X5 X6 X8  X10  ...  X375  X376  X377  X378  \
0        1  az   v   n  f  d   t  a  w    0  ...     0     0     0     1   
1        2   t   b  ai  a  d   b  g  y    0  ...     0     0     1     0   
2        3  az   v  as  f  d   a  j  j    0  ...     0     0     0     1   
3        4  az   l   n  f  d   z  l  n    0  ...     0     0     0     1   
4        5   w   s  as  c  d   y  i  m    0  ...     1     0     0     0   
...    ...  ..  ..  .. .. ..  .. .. ..  ...  ...   ...   ...   ...   ...   
4204  8410  aj   h  as  f  d  aa  j  e    0  ...     0     0     0     0   
4205  8411   t  aa  ai  d  d  aa  j  y    0  ...     0     1     0     0   
4206  8413   y   v  as  f  d  aa  d  w    0  ...     0     0     0     0   
4207  8414  ak   v  as  a  d  aa  c  q    0  ...     0     0     1     0   
4208  8416   t  aa  ai  c  d  aa  g  r    0  ...     1     0     0     0   

      X379  X380  X382  X383  X384  X385  
0        0     0     0     0     0     0  
1        0     0     0     0     0     0  
2        0     0     0     0     0     0  
3        0     0     0     0     0     0  
4        0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...  
4204     0     0     0     0     0     0  
4205     0     0     0     0     0     0  
4206     0     0     0     0     0     0  
4207     0     0     0     0     0     0  
4208     0     0     0     0     0     0  

[4209 rows x 377 columns]

In [4]:
drop=(X.var()==0).index
drop

C:\Users\Arbaj\AppData\Local\Temp/ipykernel_19020/3150818988.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  drop=(X.var()==0).index


Index(['ID', 'y', 'X10', 'X11', 'X12', 'X13', 'X14', 'X15', 'X16', 'X17',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=370)

In [5]:
X.drop(drop,axis=1,inplace=True)
X

X0 X1  X2 X3 X4  X5 X6 X8
0      k  v  at  a  d   u  j  o
1      k  t  av  e  d   y  l  o
2     az  w   n  c  d   x  j  x
3     az  t   n  f  d   x  l  e
4     az  v   n  f  d   h  d  n
...   .. ..  .. .. ..  .. .. ..
4204  ak  s  as  c  d  aa  d  q
4205   j  o   t  d  d  aa  h  h
4206  ak  v   r  a  d  aa  g  e
4207  al  r   e  f  d  aa  l  u
4208   z  r  ae  c  d  aa  g  w

[4209 rows x 8 columns]

In [6]:
len(X.columns)

8

In [7]:
X.isna().sum()

X0    0
X1    0
X2    0
X3    0
X4    0
X5    0
X6    0
X8    0
dtype: int64

In [8]:
c=X.columns
for i in c:
    A=(X[i].unique(),len(X[i].unique()))
    print(A)

(array(['k', 'az', 't', 'al', 'o', 'w', 'j', 'h', 's', 'n', 'ay', 'f', 'x',
       'y', 'aj', 'ak', 'am', 'z', 'q', 'at', 'ap', 'v', 'af', 'a', 'e',
       'ai', 'd', 'aq', 'c', 'aa', 'ba', 'as', 'i', 'r', 'b', 'ax', 'bc',
       'u', 'ad', 'au', 'm', 'l', 'aw', 'ao', 'ac', 'g', 'ab'],
      dtype=object), 47)
(array(['v', 't', 'w', 'b', 'r', 'l', 's', 'aa', 'c', 'a', 'e', 'h', 'z',
       'j', 'o', 'u', 'p', 'n', 'i', 'y', 'd', 'f', 'm', 'k', 'g', 'q',
       'ab'], dtype=object), 27)
(array(['at', 'av', 'n', 'e', 'as', 'aq', 'r', 'ai', 'ak', 'm', 'a', 'k',
       'ae', 's', 'f', 'd', 'ag', 'ay', 'ac', 'ap', 'g', 'i', 'aw', 'y',
       'b', 'ao', 'al', 'h', 'x', 'au', 't', 'an', 'z', 'ah', 'p', 'am',
       'j', 'q', 'af', 'l', 'aa', 'c', 'o', 'ar'], dtype=object), 44)
(array(['a', 'e', 'c', 'f', 'd', 'b', 'g'], dtype=object), 7)
(array(['d', 'b', 'c', 'a'], dtype=object), 4)
(array(['u', 'y', 'x', 'h', 'g', 'f', 'j', 'i', 'd', 'c', 'af', 'ag', 'ab',
       'ac', 'ad', 'ae', 'ah', 'l'

In [9]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
for i in X.columns:
    X[i]=le.fit_transform(X[i])


In [10]:
X

X0  X1  X2  X3  X4  X5  X6  X8
0     32  23  17   0   3  24   9  14
1     32  21  19   4   3  28  11  14
2     20  24  34   2   3  27   9  23
3     20  21  34   5   3  27  11   4
4     20  23  34   5   3  12   3  13
...   ..  ..  ..  ..  ..  ..  ..  ..
4204   8  20  16   2   3   0   3  16
4205  31  16  40   3   3   0   7   7
4206   8  23  38   0   3   0   6   4
4207   9  19  25   5   3   0  11  20
4208  46  19   3   2   3   0   6  22

[4209 rows x 8 columns]

In [11]:
from sklearn.decomposition import PCA
model=PCA(n_components=0.87)
model.fit(X)

PCA(n_components=0.87)

In [12]:
X=model.transform(X)

In [13]:
X.shape

(4209, 4)

In [14]:
!pip install xgboost
from xgboost import XGBRegressor
import xgboost as xgb
import time
xgm=xgb.XGBRegressor()
start=time.time()
xgm.fit(X,y)
end=time.time()
time_req=end-start
print(time_req)

1.380786418914795


In [15]:
Y

ID  X0  X1  X2 X3 X4  X5 X6 X8  X10  ...  X375  X376  X377  X378  \
0        1  az   v   n  f  d   t  a  w    0  ...     0     0     0     1   
1        2   t   b  ai  a  d   b  g  y    0  ...     0     0     1     0   
2        3  az   v  as  f  d   a  j  j    0  ...     0     0     0     1   
3        4  az   l   n  f  d   z  l  n    0  ...     0     0     0     1   
4        5   w   s  as  c  d   y  i  m    0  ...     1     0     0     0   
...    ...  ..  ..  .. .. ..  .. .. ..  ...  ...   ...   ...   ...   ...   
4204  8410  aj   h  as  f  d  aa  j  e    0  ...     0     0     0     0   
4205  8411   t  aa  ai  d  d  aa  j  y    0  ...     0     1     0     0   
4206  8413   y   v  as  f  d  aa  d  w    0  ...     0     0     0     0   
4207  8414  ak   v  as  a  d  aa  c  q    0  ...     0     0     1     0   
4208  8416   t  aa  ai  c  d  aa  g  r    0  ...     1     0     0     0   

      X379  X380  X382  X383  X384  X385  
0        0     0     0     0     0     0  
1        0     0     0     0     0     0  
2        0     0     0     0     0     0  
3        0     0     0     0     0     0  
4        0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...  
4204     0     0     0     0     0     0  
4205     0     0     0     0     0     0  
4206     0     0     0     0     0     0  
4207     0     0     0     0     0     0  
4208     0     0     0     0     0     0  

[4209 rows x 377 columns]

In [16]:
drop1=(Y.var()==0).index
drop1

C:\Users\Arbaj\AppData\Local\Temp/ipykernel_19020/700550590.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  drop1=(Y.var()==0).index


Index(['ID', 'X10', 'X11', 'X12', 'X13', 'X14', 'X15', 'X16', 'X17', 'X18',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=369)

In [17]:
Y.drop(drop1,axis=1,inplace=True)

In [18]:
Y

X0  X1  X2 X3 X4  X5 X6 X8
0     az   v   n  f  d   t  a  w
1      t   b  ai  a  d   b  g  y
2     az   v  as  f  d   a  j  j
3     az   l   n  f  d   z  l  n
4      w   s  as  c  d   y  i  m
...   ..  ..  .. .. ..  .. .. ..
4204  aj   h  as  f  d  aa  j  e
4205   t  aa  ai  d  d  aa  j  y
4206   y   v  as  f  d  aa  d  w
4207  ak   v  as  a  d  aa  c  q
4208   t  aa  ai  c  d  aa  g  r

[4209 rows x 8 columns]

In [19]:
from sklearn.preprocessing import LabelEncoder
le1=LabelEncoder()
for i in Y.columns:
    Y[i]=le1.fit_transform(Y[i])

In [20]:
Y

X0  X1  X2  X3  X4  X5  X6  X8
0     21  23  34   5   3  26   0  22
1     42   3   8   0   3   9   6  24
2     21  23  17   5   3   0   9   9
3     21  13  34   5   3  31  11  13
4     45  20  17   2   3  30   8  12
...   ..  ..  ..  ..  ..  ..  ..  ..
4204   6   9  17   5   3   1   9   4
4205  42   1   8   3   3   1   9  24
4206  47  23  17   5   3   1   3  22
4207   7  23  17   0   3   1   2  16
4208  42   1   8   2   3   1   6  17

[4209 rows x 8 columns]

In [21]:
from sklearn.decomposition import PCA
model=PCA(n_components=0.87)
model.fit(Y)

PCA(n_components=0.87)

In [22]:
Y=model.transform(Y)

In [23]:
Y

array([[ 14.4804318 ,  14.07812797,  13.47617281,   2.57844247],
       [-15.09247681,  -7.80301904,  -7.33561239,  -2.59310184],
       [ 11.77201271,  -1.71391841, -10.23300463,  14.75453317],
       ...,
       [-13.510305  ,   3.35512837,  -7.17273509,  18.74427103],
       [ 24.84396683,  -4.87308331, -10.32829395,  11.3029468 ],
       [-15.24207335,  -7.75858586, -15.47965692,  -0.82253574]])

In [24]:
import sys
import numpy as np
np.set_printoptions(threshold=sys.maxsize)

In [25]:
prediction=xgm.predict(Y)
prediction[:50]

array([ 77.10393 ,  92.338554,  94.20674 ,  86.98323 , 111.81286 ,
        92.88999 ,  94.476875,  98.896515, 118.25527 , 102.226685,
       117.517876, 110.56026 ,  98.313324,  95.45693 , 104.925896,
        99.02465 , 117.65208 ,  99.19314 , 110.336716, 106.29868 ,
        99.19314 ,  99.19314 ,  93.069305,  99.68997 ,  91.53148 ,
       110.58429 , 102.44439 , 105.08726 ,  94.903854,  95.790596,
       109.71577 , 102.04968 , 107.00115 , 101.062035, 101.84487 ,
       101.0388  ,  92.07389 , 105.28472 ,  92.402985,  92.98583 ,
       101.8221  ,  98.29079 , 100.23383 , 108.16576 ,  94.43146 ,
        96.707565, 105.28472 ,  98.77961 ,  97.481155, 108.1366  ],
      dtype=float32)

# test the model on test data for checking accuracy

In [26]:
train=xgm.predict(X)


In [27]:
from sklearn.metrics import mean_squared_error
MSE_TRAIN=mean_squared_error(y.values,train)
MSE_TRAIN

28.153700142897655

In [28]:
RMSE_TRAIN=np.sqrt(MSE_TRAIN)
RMSE_TRAIN

5.30600604437063

In [29]:
R_SQUARE=xgm.score(X,y)
R_SQUARE

0.8248369285118444

In [30]:
Y=pd.read_csv(r"C:\Users\Arbaj\Downloads\test\test.csv")
Result=Y[:50]


In [31]:
Answer=prediction[:50]

In [32]:
Result['predictions']=Answer
Result[:10]

C:\Users\Arbaj\AppData\Local\Temp/ipykernel_19020/2900382545.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Result['predictions']=Answer


ID  X0  X1  X2 X3 X4 X5 X6 X8  X10  ...  X376  X377  X378  X379  X380  \
0   1  az   v   n  f  d  t  a  w    0  ...     0     0     1     0     0   
1   2   t   b  ai  a  d  b  g  y    0  ...     0     1     0     0     0   
2   3  az   v  as  f  d  a  j  j    0  ...     0     0     1     0     0   
3   4  az   l   n  f  d  z  l  n    0  ...     0     0     1     0     0   
4   5   w   s  as  c  d  y  i  m    0  ...     0     0     0     0     0   
5   8   y  aa  ai  e  d  x  g  s    0  ...     0     0     0     0     0   
6  10   x   b  ae  d  d  x  d  y    0  ...     0     0     0     0     1   
7  11   f   s  ae  c  d  h  d  a    0  ...     0     1     0     0     0   
8  12  ap   l   s  c  d  h  j  n    0  ...     0     0     0     0     0   
9  14   o   v  as  f  d  g  f  v    0  ...     0     0     0     0     0   

   X382  X383  X384  X385  predictions  
0     0     0     0     0    77.103928  
1     0     0     0     0    92.338554  
2     0     0     0     0    94.206741  
3     0     0     0     0    86.983231  
4     0     0     0     0   111.812859  
5     0     0     0     0    92.889992  
6     0     0     0     0    94.476875  
7     0     0     0     0    98.896515  
8     0     0     0     0   118.255272  
9     0     0     0     0   102.226685  

[10 rows x 378 columns]

In [33]:
Result_sorted = Result.sort_values(by=['predictions'], ascending=True)
Result_sorted[:1]

ID  X0 X1 X2 X3 X4 X5 X6 X8  X10  ...  X376  X377  X378  X379  X380  X382  \
0   1  az  v  n  f  d  t  a  w    0  ...     0     0     1     0     0     0   

   X383  X384  X385  predictions  
0     0     0     0    77.103928  

[1 rows x 378 columns]

# RESULT

 FOR THIS DATA THE PREDICTION WILL BE SHOWN ABOVE WHICH IS 82% CORRCT 
 
 ON BASIS IF THIS THE COMPANY ENGINEER GET IDEA ABOUT REDUCING THE TIME SPENT ON PRODUCTION AND WHICH COMBINSTION SUITS BETTER FOR TIME REDUCEMENT.

Above adjustments will take minimum time for packaging unit 
It will take 77.10 seconds. 

# CONCLUSION

The model that i build is 82% accurate 

this explains the correction in prediction

the provided data will take how much of time will be predicted by the model.

time for packaging can be reduced by changing algorithms and this will improve productivity of company.

from the entire adjustments datasets the above 1 adjustment will take less time than other .